### ISSUE TITLE
Visual glitches in the Termux app for Android when running v2021.08.28, even on a linux server

### ISSUE THREAD

**Issue by buttaface** [mawww/kakoune#4339]
### Steps 
Open Kakoune with most any file in the Termux app, even when running Kakoune in Fedora linux remotely and accessing it through ssh.
 
### Outcome 

Random lines are highlighted in white, which comes and goes when the screen updates because of a keyboard press or any other touch update.

### Expected 
No random highlighting in white, see example below of running Fedora's kakoune system package in ssh from my Android 6 tablet.

![Screenshot_2021-09-10-14-17-19](https://user-images.githubusercontent.com/48358669/132827844-e3cc0cfe-4bc6-4976-a7a3-5fb1f69ecff6.png)

I'm able to reproduce on another device running Android 11, both running the Termux package locally or the Fedora package remotely through ssh.

**Comment by lenormf** [comment]
What version of Kakoune is this? Have you tried other SSH clients like JuiceSSH?

**Comment by Screwtapello** [comment]
I installed Kakoune v2012.08.28 from Termux 0.117, launched `kak -n .bashrc` and scrolled around, and did not see any white flashes like those depicted. I also tried SSHing to a Debian box and ran Kakoune inside tmux, and I didn't see any white flashes that way either.

**Comment by Screwtapello** [comment]
OK, I don't get white glitches with `kak -n` or with my normal colour scheme, but I *do* get them with `:colorscheme default`. Also, when an infobox appears, often there are stripes of yellow to the left-hand edge of the screen. Perhaps the Termux terminal emulator doesn't treat background-color-erase the way Kakoune expects?

**Comment by buttaface** [comment]
> What version of Kakoune is this? 

It's noted in the issue title, no problem with the previous release.

> Have you tried other SSH clients like JuiceSSH?

No, I only use Termux.

I'm not familiar with how the Termux app supports terminal sequences, maybe @agnostic-apollo can fill you in.

**Comment by agnostic-apollo** [comment]
I haven't really looked into the deep terminal level workings, but those commands are handled by `TerminalEmulator` and seems like resetting of background color is done [here](https://github.com/termux/termux-app/blob/v0.117/terminal-emulator/src/main/java/com/termux/terminal/TerminalEmulator.java#L2016).

It would probably require checking which commands are being sent during updates and how termux is handling them to figure out the causes for various problems above.

**Comment by Screwtapello** [comment]
I ran `script kak -n` to produce a typescript file containing all the sequences output by Kakoune. I then edited the file, removing chunks and catting it in both GNOME Terminal and Termux to isolate a difference, and I found one pretty quickly.

Terminal formatting sequences start with `ESC [` and finish with `m`, and in between is a semicolon-separated list of codes to be applied. For example, `ESC [1m` produces bold text, `ESC [32m` produces green text, `ESC [1;32m` produces bold green text, and so forth. The special code 0 means "reset everything to defaults", so while `ESC [1m` adds bold to whatever formatting was currently in effect, `ESC [0;1m` always produces a bold version of the default text style.

An additional quirk: if a terminal sequence *doesn't have a number*, there's usually a sensible default. For example, `ESC [5B` moves the cursor down five lines, but `ESC [B` moves the cursor down one line. The default for the set-formatting code is 0, so `ESC [m` resets all formatting.

So far, Termux and GNOME Terminal are in agreement. Here's the difference:

    printf '\e[4munderline\e[;3mitalic\n'

On GNOME Terminal, this produces the word "underline" underlined, followed by the word "italic" in italic.

On Termux, this produces "underline" underlined, followed by "italic" in italic *and underlined*.

That is, GNOME Terminal parses `ESC [;3m` as `ESC [0;3m` (expanding an empty field to the default value), while Termux parses it as `ESC [3m` (since a parameter exists, it doesn't apply the default). Kakoune uses `ESC [;...m` sequences all the time, so it's not surprising to me that things get confused on Termux. Termux should probably fix its formatting sequence parsing to match other terminals, but I think this is also a bug in Kakoune:

https://github.com/mawww/kakoune/blob/b3a1017a264cfea7e9ae1a374208e9b02efbda5a/src/terminal_ui.cc#L241-L277

When producing a formatting escape sequence, Kakoune uses a `join` flag so it can avoid starting an escape sequence with a semicolon, and seems to have logic to layer formatting styles together instead of deliberately resetting them from scratch every time. However, when setting a new formatting attribute, it always writes a leading semicolon regardless of the state of the `join` flag. On the other hand, I can't seem to trigger the bug I think should be there, and trying to fix it makes all the formatting go crazy. I'm not sure what's going on here.
